Add /surveys POST endpoint

dinahshi opened the pull request:
Fixes https://github.com/uwblueprint/somo/issues/60

dinahshi committed:
Change response choice to expect lower case keys

dinahshi committed:
Add number to question model

dinahshi committed:
Merge branch 'master' into post_survey

dinahshi committed:
Allow spec :focus tag

dinahshi committed:
Run all tests when everything filtered

dinahshi committed:
Add question_order#type to support single table inheritance

dinahshi committed:
Change question to have many question orders

dinahshi committed:
Add test for question order with null next_question

dinahshi committed:
Add /surveys create (POST)

chrisjluc commented on app/controllers/surveys_controller.rb:
```diff
@@ -1,4 +1,58 @@
   end
+
+  def create
+    @survey = Survey.create!(survey_params)
```
Why is survey an instance variable?

dinahshi replied:
This is a convention that allows the view to access the data which is useful when we eventually want the view to render the survey object as JSON.

chrisjluc replied:
So is it convention when in the controller and we manipulate or access a resource for that controller, we always put it in an ivar? i.e. `send_survey` method, i'll put survey into `@survey`

dinahshi replied:
Yep! Then rails will auto inject into the view and you can access it there.

chrisjluc commented on app/controllers/surveys_controller.rb:
```diff
@@ -1,4 +1,58 @@
+    @survey = Survey.create!(survey_params)
+    questions = []
+
+    params[:questions].each do |question_params|
```
What if you used `map` to build the questions array? Then you wouldn't need to push question on each iteration. I just get a cringing feeling everytime I see push on to a dynamic array


i.e.
```
questions = params[:questions].map do |question|
  question_params[:options].each do |response_params|
    ResponseChoice.Create...
  question = Question.create...
```

dinahshi replied:
Good feedback! Fixed

chrisjluc commented on app/controllers/surveys_controller.rb:
```diff
@@ -1,4 +1,58 @@
+      end
+    end
+
+    # TODO: remove this empty json when survey jbuilder is complete
```
Lets assign ownership to TODOs and also allows us to contact the creator of that TODO in the future if it's been forgotten
i.e. `TODO (Dinah): ...`

dinahshi replied:
Fixed

chrisjluc commented on app/models/question.rb:
```diff
@@ -7,15 +7,17 @@
+  validates :number, presence: true, numericality: { only_integer: true }
 
-  validates :question_type, presence: true, inclusion: {in: %w[short_answer mc checkbox t_or_f]}
+  validates :question_type, presence: true, inclusion: {in: %w[short_answer multiple_choice true_false]}
```
What happened to `checkbox`?

dinahshi replied:
Removed because I didn't see it as an option in the API doc from frontend. Given how easy it is to add in at a later date, I think we should leave it out if we're uncertain for now.

chrisjluc replied:
Fair enough.

chrisjluc commented on config/routes.rb:
```diff
@@ -6,6 +6,7 @@
   post 'sms/received_user_response' => :received_user_response, controller: 'sms'
 
   # Survey Routes
+  resources :surveys, only: [:create]
```
Can we combine this and the route below into one block

i.e.
```
resources :events, only: [:index, :show] do
    resources :registrations, except: :index
end
```

dinahshi replied:
Fixed

chrisjluc commented on app/controllers/surveys_controller.rb:
```diff
@@ -1,4 +1,58 @@
+        question_type: question_params[:question_type].downcase,
+        number: question_params[:id],
+      )
+      questions.push(question)
```
Do we have any error handling for if the question model or any other model isn't valid?

chrisjluc commented on app/controllers/surveys_controller.rb:
```diff
@@ -1,4 +1,58 @@
+            survey: @survey,
+            question: questions[question_params[:id].to_i - 1],
+            response_choice: questions[question_params[:id].to_i - 1].response_choices.find{ |r| r.key = response_params[:key] },
+            next_question: response_params[:next_question_id].to_i > 0 ? questions[response_params[:next_question_id].to_i - 1] : nil
```
Can we encapsulate `question_params[:id].to_i -1` and `response_params[:next_question_id].to_i` into a variable since you call it twice, makes it hard to read and the line is too long

chrisjluc replied:
You can even encapsulate the question instance if you wanted.

chrisjluc commented on app/controllers/surveys_controller.rb:
```diff
@@ -1,4 +1,58 @@
+        DefaultQuestionOrder.create!(
+          survey: @survey,
+          question: questions[question_params[:id].to_i - 1 ],
+          next_question: question_params[:default_next_question_id].to_i > 0 ? questions[question_params[:default_next_question_id].to_i - 1] : nil
```
Can we encapsulate as a variable

dinahshi replied:
Fixed

chrisjluc commented on app/controllers/surveys_controller.rb:
```diff
@@ -1,4 +1,58 @@
+          ConditionalQuestionOrder.create!(
+            survey: @survey,
+            question: questions[question_params[:id].to_i - 1],
+            response_choice: questions[question_params[:id].to_i - 1].response_choices.find{ |r| r.key = response_params[:key] },
```
Could the find have been done like `response_choices.find(key: response_params[:key])` ?
I'm just curious why it's in a block? and what's the difference?

dinahshi replied:
No particular reason for using a block. `find` is for ids but I can change this to `find_by_key`.

chrisjluc replied:
I find `find_by_key` easier to read.

chrisjluc commented on app/models/question.rb:
```diff
@@ -7,15 +7,17 @@
 #  created_at                           :datetime         not null
 #  updated_at                           :datetime         not null
 #  question_type                        :string
+#  number                               :integer
```
Can we move `question_type` and `number` above `created_at` since they're more important

dinahshi replied:
Fixed

chrisjluc commented on db/migrate/20161118210435_add_question_number.rb:
```diff
@@ -0,0 +1,6 @@
+class AddQuestionNumber < ActiveRecord::Migration
+  def change
+    add_column :questions, :number, :integer, unique: true
```
Does this handle the case where `number` can appear multiple times as long as it's unique within a survey? That was my understanding for number.

dinahshi replied:
Right - my bad this is the wrong place to put a uniqueness constraint!

dinahshi replied:
Now that I look at the schema, there's not an easy way to put a unique constraint on (survey_id, number) because the question table doesn't store survey_id. Will remove for now and think about this some more.

chrisjluc replied:
Why does `question_order` have `survey_id` again? 

What if `survey_id` was on `question` instead and would it be able to provide the same functionality? If we could, then we'd be able to apply that unique constraint because they'd both exist in the `question` model.

dinahshi replied:
We have this in ActiveRecord models from question `has_one :survey` but not in the tables. I think we should keep the `survey_id` on the `question_orders` table because question orders is what connects surveys to questions. We could duplicate the column across two tables but I think it's better to add a more sophisticated validation method on the survey ActiveRecord model. Again, this is more of a nice to have than necessity so I'd be willing to merge without this.

chrisjluc commented on spec/spec_helper.rb:
```diff
@@ -21,6 +21,13 @@
+    config.filter_run focus: true
+    config.run_all_when_everything_filtered = true
+  end
+
```
Can we have comments here? Not sure what's going on.

dinahshi replied:
I'll add a comment! It allows you to filter which specs you run which is helpful when you want to isolate one specific test case. This might help: https://www.relishapp.com/rspec/rspec-core/v/2-0/docs/filtering/inclusion-filters

chrisjluc replied:
oh nice! yeah this is useful.

chrisjluc commented on spec/controllers/surveys_controller_spec.rb:
```diff
@@ -2,4 +2,95 @@
 
 describe SurveysController, type: :controller do
 
+  describe "GET #create" do
```
Isn't this a POST?

dinahshi replied:
Fixed

chrisjluc commented on db/migrate/20161122035953_question_order_type.rb:
```diff
@@ -0,0 +1,5 @@
+class QuestionOrderType < ActiveRecord::Migration
+  def change
+    add_column :question_orders, :type, :string, null: false
```
What is `type` in `question_order` used for? I don't see it being referenced in this PR

dinahshi replied:
Type is what tells rails that we're using single table inheritance (i.e. DefaultQuestionOrder and ConditionalQuestionOrder being written to same table). It lets rails know which model to instantiate when it reads the row back from the db. This is more of a piece that we left out when building the question order models.

chrisjluc commented on app/controllers/surveys_controller.rb:
```diff
@@ -1,4 +1,58 @@
+      else
+        DefaultQuestionOrder.create!(
+          survey: @survey,
+          question: questions[question_params[:id].to_i - 1 ],
```
Nit: remove space at the end before the closing bracket

dinahshi replied:
Fixed

chrisjluc commented:
Sorry for all the comments! 
They're mainly just questions and minor issues that i'll leave to your discretion.

Nice PR! 👍

dinahshi committed:
Add pry to help with debugging

dinahshi committed:
Associate response choice to question

dinahshi committed:
Add comment to explain inclusion filtering

dinahshi committed:
Remove unique from question number column

Note this doesn’t require a separate migration because the original
migration was wrong. The unique constraint should have been on the
index so this was essentially useless.

dinahshi committed:
Reorder question model comments

dinahshi committed:
Change find block to find_by helper

dinahshi committed:
Refactor surveys_controller

dinahshi committed:
Consolidate surveys routes

dinahshi committed:
Comment surveys controller tests for get, put

dinahshi committed:
Merge branch 'master' into post_survey

dinahshi committed:
Add specs for error status codes

dinahshi committed:
Add base controller to catch and format errors

dinahshi committed:
Do not create question_order for last question, update specs

dinahshi commented:
Ready for another round of review. Some of the bigger changes: add `BaseController` to rescue from common errors and render the error as JSON, do not create a `QuestionOrder` for last question(s).

dinahshi committed:
Add questions association to survey model

chrisjluc commented on app/controllers/base_controller.rb:
```diff
@@ -0,0 +1,15 @@
+
+  def format_error(e)
+    {
+      error_type: e.class.to_s.underscore.split("/").last,
```
Nit: single quotes as it says in our style guide.

chrisjluc replied:
Nit: single quotes as it says in our style guide.

chrisjluc commented on app/controllers/surveys_controller.rb:
```diff
@@ -31,4 +31,67 @@ def send_survey
+    end
+    questions = questions.to_h
+
+    @survey.update!(first_question: questions["1"])
```
Nit: single quotes

chrisjluc commented on spec/controllers/surveys_controller_spec.rb:
```diff
@@ -73,4 +72,122 @@
     end
   end
+
+  describe "POST #create" do
```
Nit: single quotes and below (or we can change to single quotes for the whole file later)

chrisjluc commented on spec/models/conditional_question_order_spec.rb:
```diff
@@ -18,4 +18,10 @@
     expect(subject.survey).to be_a(Survey)
   end
+
+  context "with no next question" do
```
Nit: single quotes

chrisjluc commented on spec/models/question_spec.rb:
```diff
@@ -16,6 +16,11 @@
     it { is_expected.not_to be_valid }
   end
 
+  context "question is missing number" do
```
Not: single quotes

chrisjluc commented on app/models/survey.rb:
```diff
@@ -14,6 +14,8 @@ class Survey < ActiveRecord::Base
   has_many :question_orders
   has_many :survey_responses
 
+  has_many :questions, through: :question_orders
```
Why did we have `survey_id` on `question_orders` model instead of `questions`? Conceptually, it makes more sense to me for `questions` to have `survey_id` directly.

I'll leave this up to you to decide.

chrisjluc commented on spec/routing/survey_routing_spec.rb:
```diff
@@ -1,25 +1,25 @@
-      :id => '1'
-    )
-  end
+  # it "get surveys routes properly" do
```
Should these still be commented out? If so can you leave a comment for why?

chrisjluc reviewed (approved):
LGTM, left a few minor comments.

dinahshi committed:
Add todo comments for commented survey route specs

dinahshi committed:
Change double quotes to single quotes

dinahshi committed:
Move survey association from question_order to question

dinahshi committed:
Removing double quotes
